github-api-call-inventory/github-api-call-inventory.ipynb

```diff
@@ -0,0 +1,720 @@
+       "grand_total = 138"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "print(\"6 chroots, config is called '.packit.yaml', no tests\")\n",
+    "github_api_calls(\n",
+    "    chroots=6, \n",
+    "    rate_limit=5000, \n",
+    "    load_packit_yaml=2+1, \n",
+    "    builds_enabled=True, \n",
+    "    tests_enabled=False\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "42af2a1a",
+   "metadata": {},
+   "source": [
+    "## Ways to improve things"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0c05d101",
+   "metadata": {},
+   "source": [
+    "The main issue seems to be that `task.steve_jobs.process_message` re-creates the context of the pipeline by getting information from GitHub: checks if the project is private, if the user is approved to run the pipeline, and load the package config (`.packit.yaml`).\n",
+    "\n",
+    "All this could be avoided by looking up the pipeline when an event arrives (Copr events, testing farm events - when these are received there already needs to be a pipeline created, otherwise they are invalid), and getting this information from the pipeline.\n",
+    "\n",
+    "The only time it makes sense to load this from GitHub is when the pipeline is started, for example when the PR event is received."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "616bb63c",
+   "metadata": {},
+   "source": [
+    "Let's pick the worst case from the list above (*6 chroots, config is called 'packit.yaml'*) as a reference, and check how the number of API calls change if we apply the above.\n",
+    "\n",
+    "**Load the package config only once per pipeline run** to increase the number of possible runs from 14 to 38 (~171% increase). This is achived by reducing the number of API calls to get the package config from 222 to 6."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "c1b001e5",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "prs_per_hour_per_org = 38\n",
+       "total_is_project_private = 37\n",
+       "total_is_user_approved = 2\n",
+       "total_load_packit_yaml = 6\n",
+       "total_set_status = 84\n",
+       "grand_total = 129"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "github_api_calls(\n",
+    "    chroots=6, \n",
+    "    rate_limit=5000,\n",
+    "    load_packit_yaml=5+1,\n",
+    "    builds_enabled=True,\n",
+    "    tests_enabled=True,\n",
+    "    load_packit_yaml_once=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2284d964",
+   "metadata": {},
+   "source": [
+    "**Don't update the test flags while the build is in progress**, to get to 47 PRs/hour/org (a 235% increase in total), thanks to shaving of 24 API calls to update the status flags."
+   ]
```

Comment:
Currently when build events are handled both build flags (ex: `rpm-build:fedora-37-x86_64`) and test flags (ex: `testing-farm:fedora-37-x86_64`) are updated.

This means that test flags will cycle through the same 

1. "SRPM build in Copr was submitted"
1. "SRPM build in Copr has started..."
1. "SRPM build in Copr has finished."
2. "RPM build is in progress..."
3. "RPMs were built successfully."

messages. Setting each of these is another API call. Instead of doing this when the SRPM build in Copr is submitted, test flags could be set to something like "Waiting for RPMs to be built...", and updated only when tests are submitted.

Reply:
Yes, that's what I meant. Thanks and agree.